src/runtime-rs/crates/shimv2/src/shim_start.rs

```diff
@@ -0,0 +1,221 @@
+
+        let address = self.do_start().context("do start")?;
+        std::io::stdout()
+            .write_all(address.as_os_str().as_bytes())
+            .context("failed to write stdout")?;
+        Ok(())
+    }
+
+    fn do_start(&mut self) -> Result<PathBuf> {
+        let spec = self.load_oci_spec()?;
+        let (container_type, id) = k8s::container_type_with_id(&spec);
+
+        match container_type {
+            ContainerType::PodSandbox => {
+                let address = self.socket_address(&self.args.id)?;
+                let socket = new_listener(&address)?;
+                let child_pid = self.create_shim_process(socket)?;
+                self.write_pid_file(child_pid)?;
+                self.write_address(&address)?;
+                Ok(address)
+            }
+            ContainerType::PodContainer => {
+                let sid = id
+                    .ok_or(Error::InvalidArgument)
+                    .context("get sid for container")?;
+                let (address, pid) = self.get_shim_info_from_sandbox(&sid)?;
+                self.write_pid_file(pid)?;
+                self.write_address(&address)?;
+                Ok(address)
+            }
+        }
+    }
+
+    fn new_command(&self) -> Result<std::process::Command> {
+        if self.args.id.is_empty()
+            || self.args.namespace.is_empty()
+            || self.args.address.is_empty()
+            || self.args.publish_binary.is_empty()
+        {
+            return Err(anyhow!("invalid param"));
+        }
+
+        let bundle_path = self.get_bundle_path()?;
+        let self_exec = std::env::current_exe().map_err(Error::SelfExec)?;
+        let mut command = std::process::Command::new(self_exec);
+
+        command
+            .current_dir(bundle_path)
+            .stdin(std::process::Stdio::null())
+            .stdout(std::process::Stdio::null())
+            .stderr(std::process::Stdio::null())
+            .arg("-id")
+            .arg(&self.args.id)
+            .arg("-namespace")
+            .arg(&self.args.namespace)
+            .arg("-address")
+            .arg(&self.args.address)
+            .arg("-publish-binary")
+            .arg(&self.args.publish_binary)
+            .env("RUST_BACKTRACE", "1");
+
+        if self.args.debug {
+            command.arg("-debug");
+        }
+
+        Ok(command)
+    }
+
+    fn create_shim_process<T: IntoRawFd>(&self, socket: T) -> Result<u32> {
+        let mut cmd = self.new_command()?;
+        cmd.env(ENV_KATA_BIND_FD, format!("{}", socket.into_raw_fd()));
+        let child = cmd.spawn().map_err(Error::SpawnChild)?;
+
+        Ok(child.id())
+    }
+
+    fn get_shim_info_from_sandbox(&self, sandbox_id: &str) -> Result<(PathBuf, u32)> {
+        // All containers of a pod share the same pod socket address.
+        let address = self.socket_address(sandbox_id)?;
+        let bundle_path = self.get_bundle_path()?;
+        let sandbox_bundle_path = bundle_path
+            .join("..")
```

Comment:
Is there any way to avoid a relative path like this?

Reply:
Thanks for the review, replace with bundle_path.parent()